Local live-live builds

jkoritzinsky opened the pull request:
This PR implements local live-live builds. The `eng/liveBuilds.targets` MSBuild file is the central location for live dependency file discovery. There are various MSBuild property hooks in there to enable overriding the location of the various consumed live components.

Work completed in this PR:

- Libraries consumes:
	- A live copy of the CoreCLR build's shared framework and System.Private.CoreLib output.
- Installer consumes:
	- A live copy of the CoreCLR build's shared framework and System.Private.CoreLib output
	- A live copy of the Libraries shared framework ref and implementation assemblies.
	- A live copy of Microsoft.NETCore.Platforms and Microsoft.NETCore.Targets
- CoreCLR managed test build consumes:
	- A live copy of the Libraries shared framework and OOB ref assemblies.
- CoreCLR Core_Root generation:
	- A live copy of the Libraries shared framework and OOB implementation assemblies.
- Customization points for each "unit" of output that is consumed between subset categories.
- Subsets that have dependencies are serialized such that a root build will build subsets in the correct order for inter-subset dependencies.

Future work for another PR once we are off the critical path:

- Refactor out target overrides in `src/installer/pkg/projects/netcoreapp/src/localnetcoreapp.override.targets` so we can send back a patch to Arcade and remove `localnetcoreapp.override.targets`.
- Extend `eng/liveBuilds.targets` to have a target for consuming a live build of the hosting layer.
- Consume a live build of the hosting layer in the CoreFX test build instead of consuming packages.

Superscedes #55 (needed to be recreated after the repo went public).

jkoritzinsky committed:
Convert libraries and installer to use a live CoreCLR.

jkoritzinsky committed:
Installer uses a live libraries build for DLLs.

jkoritzinsky committed:
Remove newline in path.

jkoritzinsky committed:
Enable live-live builds to build required live dependencies when the dependencies are in the building subset/subset category.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Update path to coreclr output.

jkoritzinsky committed:
Fix typo in coreclr.proj.

jkoritzinsky committed:
Remove last usage of CoreCLRPDBOverridePath.

jkoritzinsky committed:
Update path of binplaced runtime to include OS and Arch.

jkoritzinsky committed:
Fix typo in coreclr/tests/src/Common/Directory.Build.targets.

jkoritzinsky committed:
Restore CoreFX shared framework from live build for CoreCLR test dependencies.

jkoritzinsky committed:
Remove dependency on Microsoft.NETCore.CoreCLR.TestDependencies package.

jkoritzinsky committed:
Enable referencing live builds of the OOB libraries by temporarily adding a new BinplaceConfiguration. Update test_dependencies in the CoreCLR tree to use that to restore OOB libraries.

jkoritzinsky committed:
Remove reference to System.Drawing.Common from NETClientPrimitives (the reference is only needed at runtime and Core_Root already has System.Drawing.Common).

jkoritzinsky committed:
Remove package-based checked CoreCLR + CoreFX test runs. In the live-live world, we'll run this test configuration by pointing the CoreFX test host build step at the artifacts from a checked CoreCLR build (possible in both local and CI builds).

jkoritzinsky committed:
Only copy the UCRT on Windows.

jkoritzinsky committed:
CoreCLR test build now builds against a live libraries sfx+oob ref assembly set.

jkoritzinsky committed:
Override GetCorePackagePaths target to not reference unused packages (breaks the live-live build on a clean system).

jkoritzinsky committed:
Remove live references to Microsoft.NETCore.Platforms now that it seems to be unneeded since the SDK now packages a RID graph. Use the live RID graph in test_runtime.csproj where we may actually be looking for the live RID graph.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Fix coreclr build arg generation bug that was causing build failures.

jkoritzinsky committed:
Remove references to Microsoft.NETCore.Platforms outside of package references from the installers.

jkoritzinsky committed:
Fix problems in installer build in live-live that were found from a clean build.

jkoritzinsky committed:
Merge branch 'master' into live-live-builds

jkoritzinsky committed:
Remove unused dependency on Microsoft.NETCore.Targets.

jkoritzinsky committed:
Fix Microsoft.NETCore.App -> Microsoft.NETCore.Platforms dependency.

jkoritzinsky committed:
Remove Microsoft.NETCore.Targets version from Versions.props.

jkoritzinsky committed:
Remove unused deps.csproj file that was the last dependency on the packaged Microsoft.NETCore.Platforms package.

jkoritzinsky committed:
Merge branch 'live-live-builds' of https://github.com/jkoritzinsky/runtime into live-live-builds

jkoritzinsky committed:
Fix OSGroup calculation for dependent projects off-Windows.

jkoritzinsky committed:
Fix construction of Microsoft.NETCore.App.Internal package.

jkoritzinsky committed:
Remove unneeded CoreRun copying.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Serialize subsets that have dependencies across subset groups. Don't use `<MSBuild />` tasks to rebuild subsets since we don't want to rebuild projects multiple times (coreclr was being built multiple times).

jkoritzinsky committed:
Cleanup from PR feedback.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Only copy libraries outputs to artifacts/bin/runtime.

jkoritzinsky committed:
Use the netcoreapp output directories for OOB-included references to libraries outputs.

jkoritzinsky committed:
Remove references to deleted old corefx override setup.

jkoritzinsky committed:
First pass to get new runtime pipeline using the local live-live infra (coreclr+libraries only for now).

jkoritzinsky committed:
Don't generate layout during test native build.

jkoritzinsky commented:
/azp run runtime

ViktorHofer commented on Directory.Build.props:
```diff
@@ -46,6 +50,12 @@
+    Get ProjectToBuild and '<subset>ProjectToBuild' items. Using the items lets projects handle
+    $(Subset) automatically when creating project-to-project dependencies.
+  -->
+  <Import Condition="'$(SubsetsPropsImported)' != 'true'" Project="$(RepositoryEngineeringDir)Subsets.props" />
```
Pls remove. Subsets are already imported in the same file in L35.

ViktorHofer commented on Directory.Build.props:
```diff
@@ -18,6 +18,10 @@
   <Import Project="Sdk.props" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(SkipImportArcadeSdkFromRoot)' != 'true'" />
 
+  <PropertyGroup>
+    <RootDirectoryBuildPropsImported>true</RootDirectoryBuildPropsImported>
```
Pls remove. We don't import Directory.Build.props anymore from Subsets.props.

ViktorHofer commented on eng/Subsets.props:
```diff
@@ -48,12 +48,15 @@
   -->
 
   <PropertyGroup>
+    <SubsetsPropsImported>true</SubsetsPropsImported>
```
As mentioned above. Pls remove.

ViktorHofer commented on eng/Subsets.props:
```diff
@@ -48,12 +48,15 @@
     <RepoTasksDir>$([MSBuild]::NormalizeDirectory('$(RepoToolsLocalDir)', 'tasks'))</RepoTasksDir>
   </PropertyGroup>
+  <!-- Import only when imported by Arcade's Build.proj as we import Directory.Build.props ourselves. -->
+  <Import Project="$(RepoRoot)Directory.Build.props" Condition="'$(SkipImportArcadeSdkFromRoot)' == 'true' and '$(RootDirectoryBuildPropsImported)' != 'true'" />
```
As mentioned above, pls remove.

ViktorHofer commented on src/coreclr/tests/src/Common/Directory.Build.targets:
```diff
@@ -16,20 +16,20 @@
-      <RunTimeDependecyExcludeFiles Include="@(RunTimeDependecyExclude -> '%(FileName)%(Extension)')" />
-      <RunTimeDependecyExcludeFiles Include="@(RunTimeDependecyExclude -> '%(FileName).ni%(Extension)')" />
-      <RunTimeDependecyExcludeFiles Include="@(RunTimeDependecyExclude -> '%(FileName).pdb')" />
+      <RunTimeDependencyExclude Include="$(CORE_ROOT)\**\*.*"  />
```
nit: `RunTime` spelling looks weird.

ViktorHofer commented on src/coreclr/tests/src/Directory.Build.targets:
```diff
@@ -199,6 +199,12 @@
   <Target Name="UpdateReferenceItems"
           BeforeTargets="BeforeResolveReferences"
   >
+   <MSBuild Projects="$(MSBuildProjectFullPath)"
```
Why can't we use normal target dependencies here?

jkoritzinsky replied:
For tests that reference System.Private.CoreLib directly and don't reference the libraries, the test build for that specific test doesn't need to depend on the outputs of the libraries build. By implementing this via a conditional `<MSBuild>` task instead of a conditional `<ItemGroup>`, we don't need a libraries build for something referencing S.P.CL directly. This isn't strictly required now, but it made more sense in an earlier iteration of this PR where the `ResolveLibrariesFromLocalBuild` target could rebuild the libraries if they were out of date.

trylek replied:
Hmm, does that mean that those tests never reference other framework libraries, e.g. System.IO or such?

jkoritzinsky replied:
Yep. They reference only `System.Private.CoreLib`. There's a small number of these tests, mainly in the Interop and AppModel space for APIs that are not exposed in any public contracts but are called directly from the hosting layer.

It also used to be used to test APIs in CoreCLR that we hadn't yet exposed via CoreFX so we could add tests to a S.P.CL API in the same PR as adding the API (once again mainly in the Interop/App Model space since many of our tests include native components or need to be in their own executable and as a result lived in the CoreCLR test tree). With live-live being enabled, this usage is no longer relevant.

trylek replied:
I see, thanks for clarification. In fact I hit trouble with these tests multiple times during the switch-over to the runtime repo and in the live-live pipeline, the special conditions in the scripts for dealing with them are just so annoying. If there's any chance to make this somewhat simpler now we're in the combined repo, I would be super-happy.

jkoritzinsky replied:
I think we've already reduced the set of tests using this functionality to as few as possible, but we can add that as another thing to check over once we get live-live working.

trylek replied:
I think we can safely ignore this for now as these tests are certain to jump up during any future script refactoring so they act as an automated work item.

ViktorHofer reviewed (approved):
In general this looks really good. Thanks a lot, I think we are close.

jkoritzinsky committed:
Fix ref assembly default binplacing for netcoreapp.

jkoritzinsky committed:
Remove msbuild goo around importing Subsets.props and Directory.Build.props only once each. This goo isn't needed anymore.

jkoritzinsky committed:
Fix disabling of binplacing for the target vertical to exclude only runtime.depproj and not exclude winrt.depproj.

jkoritzinsky committed:
Fix libraries test build.

jkoritzinsky committed:
Copy all paths used by the "ResolveLibrariesFromLocalBuild" target to publish from libraries build.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

safern commented on eng/pipelines/libraries/build-job.yml:
```diff
@@ -107,19 +107,43 @@ jobs:
+          - task: CopyFiles@2
+            displayName: Prepare shared framework implementation assemblies to publish
+            inputs:
+              sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/pkg/netcoreapp/runtime
```
What does this path contain?

safern replied:
Ah sorry. It is the runtime for the runtime Microsoft.NETCore.App package.

jkoritzinsky committed:
For shims, reference the System.Private.CoreLib directly from CoreCLR outputs instead of through runtime.depproj.

jkoritzinsky committed:
Hard-code the libraries configuration to release in coreclr test builds.

jkoritzinsky committed:
Resolve ReferenceFromRuntime items directly from the CoreCLR output instead of via the runtime.depproj project.

jkoritzinsky commented:
/azp run runtime

jkoritzinsky committed:
Fix test host generation. Fix running the XML Serialization generator as part of libraries test build.

jkoritzinsky committed:
Fix XmlSerializer.Generator.Tests again.

jkoritzinsky committed:
Don't look for CoreCLR files when using ReferenceFromRuntime and not using netcoreapp.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Crossgen-comparison should depend on a libraries build so we can populate Core_Root.

jkoritzinsky committed:
Add job dependency on libraries build from run-test-job to ensure we wait correctly. Otherwise we might have a race since the managed test build might depend on one libraries build while the test run depends on another.

jashook reviewed (approved):
lgtm thank you for the work!

jkoritzinsky committed:
Only resolve coreclr files from local build when targetting netcoreapp.

jkoritzinsky committed:
Add steps to download libraries build for crossgen-comparison.

jkoritzinsky committed:
Revive old method of resolving references to the runtime for building against old netcoreapp (ie netcoreapp 3.0). Fix correctly passing along aliases info.

jkoritzinsky committed:
Fix failing AllConfigurations leg that was due to the netcoreapp runtime.depproj getting binplaced in the runtime output folder.

jkoritzinsky commented:
@safern @dagood @trylek can you take a review pass when you have a chance? This PR should be feature-complete per our agreement that breaking the `runtime-installer` pipeline is OK.

dagood commented on eng/Subsets.props:
```diff
@@ -119,23 +121,23 @@
 
   <ItemGroup Condition="$(_subsetCategory.Contains('installer')) and $(_subset.Contains('pkgproj'))">
-    <PkgprojProjectToBuild Include="$(InstallerProjectRoot)pkg\projects\**\*.pkgproj" SignPhase="MsiFiles" />
+    <PkgprojProjectToBuild Include="$(InstallerProjectRoot)pkg\projects\**\*.pkgproj" SignPhase="MsiFiles" BuildInParallel="false" />
```
Does `BuildInParallel="false"` mean every matched project runs in sequence within the group? Or can these still run in parallel and it only affects overall orchestration?

jkoritzinsky replied:
The `BuildInParallel="false"` makes all of the projects run sequentially. We'd need a fix for dotnet/arcade#4436 before we can do something more in-depth.

dagood replied:
Well, I argue in https://github.com/dotnet/arcade/issues/4436 we don't need any arcade change to do this "properly". I don't object to merging this workaround to get the PR through, though.

jkoritzinsky replied:
Theoretically I agree with you, but in practice I spent a whole day on it and couldn't get it to work successfully with coreclr's current build infrastructure.

dagood replied:
Yeah, if I notice it affecting my dev loop perf I'll file something and maybe give it a try, until then I'm fine just having https://github.com/dotnet/arcade/issues/4436 track this in case arcade can end up helping.

trylek commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,95 @@
+
+  <PropertyGroup>
+    <HostArch>$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant)</HostArch>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)' == ''">$(OSGroup)</CoreCLROSGroup>
```
How are you dealing with OS subgroup?

jkoritzinsky replied:
Currently, the local output directories in the CoreCLR product build don't include the OS subgroup IIRC, so I didn't need to do anything special here to handle it.

trylek replied:
Ahh, right you are, they don't.

dagood commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,95 @@
+<Project>
+
+  <PropertyGroup>
+    <HostArch>$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant)</HostArch>
```
This is also called `BuildArchitecture`, why can't that be shared more broadly, or have one use the value of the other?

https://github.com/dotnet/runtime/blob/d2685d5c237ba0e4d16f94b3e918c627c5f26dbd/src/installer/Directory.Build.props#L48

`ToLowerInvariant` is also missing `()` in this PR's line, but I suppose I don't know if that's actually necessary in MSBuild.

dagood commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,95 @@
+      <CoreCLRArtifactsPath>$([MSBuild]::NormalizeDirectory('$(CoreCLRArtifactsPath)'))</CoreCLRArtifactsPath>
+      <CoreCLRSharedFrameworkDir>$([MSBuild]::NormalizeDirectory('$(CoreCLRArtifactsPath)','sharedFramework'))</CoreCLRSharedFrameworkDir>
+      <CoreCLRCrossTargetComponentDir
+        Condition="'$(CoreCLRCrossTargetComponentDirName)' != ''">$([MSBuild]::NormalizeDirectory('$(CoreCLRArtifactsPath)','$(CoreCLRCrossTargetComponentDir)','sharedFramework'))</CoreCLRCrossTargetComponentDir>
```
```suggestion
        Condition="'$(CoreCLRCrossTargetComponentDirName)' != ''">$([MSBuild]::NormalizeDirectory('$(CoreCLRArtifactsPath)','$(CoreCLRCrossTargetComponentDirName)','sharedFramework'))</CoreCLRCrossTargetComponentDir>
```

However I think this file will generally need to be updated vs. https://github.com/dotnet/runtime/pull/705 so I'm not reading this file in detail except things that stand out as new.

trylek commented on eng/pipelines/coreclr/templates/crossgen-comparison-job.yml:
```diff
@@ -10,6 +10,13 @@ parameters:
+  # a dependency of this job on the appropriate libraries build and is used
+  # to construct the name of the Azure artifact representing libraries build
+  # to use for building the tests.
+  liveLibrariesBuildConfig: ''
```
I see you've also made the live-live crossgen comparison change. For now I assume that you're going to merge it in together with your other changes unless you tell me otherwise.

dagood commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,95 @@
+      <LibrariesRuntimeFiles Condition="'$(IncludeOOBLibraries)' == 'true'" Include="
+        $(LibrariesAllBinArtifactsPath)*.dll;
+        $(LibrariesAllBinArtifactsPath)*.pdb"
+        IsNative="" />
```
What does setting it to empty do?

jkoritzinsky replied:
Setting it to empty makes the metadata present so we don't have to qualify usage with `%(LibrariesRuntimeFiles.IsNative)`, which has been a constant source of typos.

trylek commented on eng/pipelines/coreclr/templates/crossgen-comparison-job.yml:
```diff
@@ -78,6 +89,16 @@ jobs:
+          artifactFileName: '$(librariesBuildArtifactName)$(archiveExtension)'
+          artifactName: '$(librariesBuildArtifactName)'
+          displayName: 'live-built libraries'
+
```
I think we'll also need $(librariesOverrideArg) below at line 116 (build-test script invocation) unless I'm mistaken.

trylek replied:
Hmm, perhaps not, looks like you've already changed this logic.

trylek commented on eng/pipelines/coreclr/templates/run-test-job.yml:
```diff
@@ -56,6 +56,8 @@ jobs:
     - ${{ if ne(parameters.stagedBuild, true) }}:
       - ${{ format('coreclr_product_build_{0}{1}_{2}_{3}', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.buildConfig) }}
+      - ${{ if ne(parameters.liveLibrariesBuildConfig, '') }}:
+        - ${{ format('libraries_build_{0}_{1}{2}_{3}_{4}', 'netcoreapp', parameters.osGroup, parameters.osSubgroup, parameters.archType, parameters.liveLibrariesBuildConfig) }}
```
This has already been merged in, I'm afraid you'll need to rebase your change.

jkoritzinsky replied:
I pulled down dotnet/runtime master and git resolved the change correctly, so I shouldn't have to rebase.

trylek commented on src/coreclr/coreclr.proj:
```diff
@@ -5,12 +5,12 @@
     <ItemGroup>
       <_CoreClrBuildArg Condition="$([MSBuild]::IsOsPlatform(Windows))" Include="-skiptests" />
-      <_CoreClrBuildArg Include="-$(Platform)" />
+      <_CoreClrBuildArg Condition="'$(TargetArchitecture)' != ''" Include="-$(TargetArchitecture)" />
```
Hmm, is there any valid scenario where we need to build CoreCLR without a target architecture specified?

jkoritzinsky replied:
We don't need to specify the arch in local builds when we're building the same architecture as we're currently running on. I can likely clean this up to always calculate the argument correctly, but I'd have to take another look (this was an early fix to get everything building).

trylek replied:
Non-blocking from my side, I was just curious. I guess that if you're in the final round of testing, you can safely ignore that for now, but I doubt so considering the current infra state :-). In fact it would be probably better to tackle in a follow-up change one way or the other, the change is already daunting w.r.t. its complexity and additional churn won't help it unless absolutely necessary for fixing severe issues, at least that's my opinion.

jkoritzinsky commented on eng/Subsets.props:
```diff
@@ -96,14 +96,16 @@
   <!-- CoreClr sets -->
   <ItemGroup Condition="$(_subsetCategory.Contains('coreclr')) and $(_subset.Contains('all'))">
-    <ProjectToBuild Include="$(CoreClrProjectRoot)coreclr.proj" />
+    <CoreClrProjectToBuild Include="$(CoreClrProjectRoot)coreclr.proj" BuildInParallel="false" />
```
Yep. Each non-parallel process builds in the order they're `Include`d. This is our current workaround for that issue.

trylek reviewed (approved):
Overall looks great, thanks Jeremy for your hard work putting this together! I tried to follow development of this change in the past weeks but given its extent I don't claim that I'm able to follow every single property and item through the script changes. I guess we'll find soon enough.

jkoritzinsky committed:
Update eng/liveBuilds.targets

Co-Authored-By: Davis Goodin <dagood@users.noreply.github.com>

jkoritzinsky committed:
HostArch->BuildArchitecture and share with installer.

jkoritzinsky committed:
Merge branch 'live-live-builds' of https://github.com/jkoritzinsky/runtime into live-live-builds

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

safern commented on Directory.Build.props:
```diff
@@ -89,6 +90,8 @@
     <!-- Used for launchSettings.json and runtime config files. -->
     <AppDesignerFolder>Properties</AppDesignerFolder>
+    <!-- Determine what architecture we are building on. -->
+    <BuildArchitecture>$([System.Runtime.InteropServices.RuntimeInformation]::ProcessArchitecture.ToString().ToLowerInvariant())</BuildArchitecture>
```
Does this dictates what architecture the product targets when building? Or can the target arch be overridden via an argument?

jkoritzinsky replied:
This is the architecture the build process is running on, separate from the target architecture.

safern commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,94 @@
+
+  <PropertyGroup>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)' == ''">$(OSGroup)</CoreCLROSGroup>
+    <CoreCLRConfiguration Condition="'$(CoreCLRConfiguration)' == ''">$(Configuration)</CoreCLRConfiguration>
```
Can I build CoreCLR for Release and then consume it in a Debug build of Libraries/installer? I presume I would need to pass these arguments down if I want to achieve that?

jkoritzinsky replied:
Yep! Just set the `CoreCLRConfiguration` or `LibrariesConfiguration` parameters to the configuration you want and it will use that configuration instead. Due to the lack of a good customization point in Arcade, the root build script will always build with the same configuration all-up, so that's why it defaults to the same config.

safern replied:
I see. It might be good if we could explore building partitions in different configurations as part of the build script. 

The reason why I mention that is because when people add an API to coreclr and then try to add tests for it in libraries, if they endup with a Debug coreclr, their tests will be super slow, so it might be good to configure it from the root script to build each partition in different configurations as one command. Rather than having to do `coreclr.cmd -Release && libraries.cmd /p:CoreCLRConfiguration=Release`.

safern commented on eng/liveBuilds.targets:
```diff
@@ -0,0 +1,94 @@
+        $(LibrariesSharedFrameworkBinArtifactsPath)*.dll;
+        $(LibrariesSharedFrameworkBinArtifactsPath)*.pdb"
+        IsNative="" />
+      <LibrariesRuntimeFiles Condition="'$(IncludeOOBLibraries)' == 'true'" Include="
```
When is this combined mode of OOB + Shared framework used?

safern replied:
I see it is used in CoreCLR tests.

jkoritzinsky replied:
This combined mode is used in the CoreCLR test build and Core_Root setup. There are various tests in CoreCLR that either reference OOB assemblies or use features of the runtime that call into OOB assemblies directly, so we need to include them.

safern commented on eng/pipelines/libraries/base-job.yml:
```diff
@@ -74,7 +74,7 @@ jobs:
           - _coreClrDownloadPath: '$(Build.SourcesDirectory)/artifacts/transport/coreclr'
           - _coreClrArtifactName: 'CoreCLRProduct_${{ parameters.osGroup }}${{ parameters.osSubgroup }}_${{ parameters.archType }}_${{ parameters.liveCoreClrBuildConfig }}'
-          - _coreClrOverridePathArg: ' /p:CoreCLROverridePath=$(_coreClrDownloadPath)'
+          - _coreClrOverridePathArg: ' /p:CoreCLRArtifactsPath=$(_coreClrDownloadPath)'
```
Nit: we can clean-up later, but it might be worth to change the variable name as well to be consistent in between MSBuild property and variable name.

safern commented on src/libraries/Directory.Build.targets:
```diff
@@ -1,4 +1,10 @@
 <Project InitialTargets="AddSkipGetTargetFrameworkToProjectReferences">
+  <PropertyGroup>
+    <ImportLiveBuildsTargets>true</ImportLiveBuildsTargets>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)'==''">$(_bc_OSGroup)</CoreCLROSGroup>
```
Nit space in between the `==` and `'`.

safern commented on src/libraries/Directory.Build.targets:
```diff
@@ -1,4 +1,10 @@
+  <PropertyGroup>
+    <ImportLiveBuildsTargets>true</ImportLiveBuildsTargets>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)'==''">$(_bc_OSGroup)</CoreCLROSGroup>
+    <CoreCLRConfiguration Condition="'$(CoreCLRConfiguration)'==''">$(_bc_ConfigurationGroup)</CoreCLRConfiguration>
```
Ditto.

safern commented on src/libraries/restore/runtime/runtime.depproj:
```diff
@@ -5,22 +5,17 @@
     <NoWarn>$(NoWarn);NU1603;NU1605</NoWarn>
     <SwapNativeForIL Condition="'$(SwapNativeForIL)' == '' and ('$(ConfigurationGroup)' == 'Debug' or '$(Coverage)' == 'true')">true</SwapNativeForIL>
+    <ImportLiveBuildsTargets>true</ImportLiveBuildsTargets>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)'==''">$(DefaultOSGroup)</CoreCLROSGroup>
```
Ditto.

safern commented on src/libraries/restore/runtime/runtime.depproj:
```diff
@@ -5,22 +5,17 @@
     <SwapNativeForIL Condition="'$(SwapNativeForIL)' == '' and ('$(ConfigurationGroup)' == 'Debug' or '$(Coverage)' == 'true')">true</SwapNativeForIL>
+    <ImportLiveBuildsTargets>true</ImportLiveBuildsTargets>
+    <CoreCLROSGroup Condition="'$(CoreCLROSGroup)'==''">$(DefaultOSGroup)</CoreCLROSGroup>
+    <CoreCLRConfiguration Condition="'$(CoreCLRConfiguration)'==''">$(ConfigurationGroup)</CoreCLRConfiguration>
```
Ditto.

safern commented on src/libraries/restore/runtime/runtime.depproj:
```diff
@@ -5,22 +5,17 @@
     <RuntimeIdentifier Condition="'$(RuntimeOS)' == 'webassembly'">$(ToolRuntimeRID)</RuntimeIdentifier>
     <NoWarn>$(NoWarn);NU1603;NU1605</NoWarn>
     <SwapNativeForIL Condition="'$(SwapNativeForIL)' == '' and ('$(ConfigurationGroup)' == 'Debug' or '$(Coverage)' == 'true')">true</SwapNativeForIL>
+    <ImportLiveBuildsTargets>true</ImportLiveBuildsTargets>
```
Why we don't import them always?

jkoritzinsky replied:
In the initial design I decided to make it opt-in, but I don't entirely know why and can't think of a good justification with the final design. I'll remove the opt-in and make it always-imported.

safern reviewed (approved):
Some questions and NITs. LGTM

jkoritzinsky committed:
Fix format nits.

jkoritzinsky committed:
Update YAML variable name.

jkoritzinsky committed:
Remove setting property to import liveBuilds.targets.

jkoritzinsky commented:
The failure in the runtime-coreclr (CoreCLR Pri0 Test Run Linux x64 checked) job looks to be a Helix infrastructure failure based on the logs for the only failed workitem:

jkoritzinsky commented:
Other than that infra failure (I'll go rerun the job to verify), this PR is ready to merge.

jkoritzinsky commented:
Now that #705 is merged, I'll get this passing for the installer legs as well.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/dotnet/runtime into live-live-builds

jkoritzinsky committed:
Upload docs and packages folders from libraries build

jkoritzinsky committed:
Inclue doc files in the Microsoft.NETCore.App.Internal package

jkoritzinsky committed:
Try making the runtime-installer pipeline live-live using the build-coreclr-and-libraries-job template.

jkoritzinsky committed:
Update artifacts download paths and msbuild properties for installer build.

jkoritzinsky committed:
Mark cross-component files as native.

jkoritzinsky committed:
Remove FrameworkPackageName to  remove error about missing $(FrameworkPackageName).versions.txt file that other shared frameworks also don't generate.

jkoritzinsky committed:
Restore GetFilesFromPackageResolve override.

jkoritzinsky commented:
I am so confused by the failures here. The failures say that the CoreCLR subset was not found at `path/to/artifacts/transport/coreclr`, but just 3 steps earlier, the CoreCLR artifacts were downloaded and extracted to that exact folder...

jkoritzinsky commented:
And with the exact same parameters it passes locally....

jkoritzinsky commented:
And the exact same thing works perfectly in the libraries builds.

jkoritzinsky committed:
Don't clean the unpack folder for the libraries download. It wipes out the CoreCLR download.

jkoritzinsky committed:
Make sure that we restore the WinRT targetting pack during the installer build.

jkoritzinsky committed:
Add libraries package directory to RestoreSources for the case where the LibrariesConfiguration is different from the installer build configuration.

jkoritzinsky committed:
Don't pass OfficialBuildId unless it is an official build. We want to be consistent with the libraries build so our package versions match.

jkoritzinsky committed:
Give default value for OfficialBuildArg variable

jkoritzinsky committed:
Ensure installer tests can see the libraries-generated packages. Fix typo in TargetPath for CoreCLR assets (runtime instead of runtimes).

Fix paths for CoreCLR cross-target files as well as crossgen to place them in the correct directories.

jkoritzinsky commented:
Everything is passing except the check for the "platform manifest" in the Microsoft.NETCore.App.Ref targeting pack. This is related to https://github.com/dotnet/core-setup/issues/8853.

ViktorHofer commented:
That's great! Does Davis's note help?

> I think I should work on this, and in the meantime I think it's reasonable to disable "full" platform manifest generation (set BuildFullPlatformManifest to false) to unblock dotnet/runtime "live live" builds.

jkoritzinsky commented:
I already followed that when I started this PR and I think that might be part of what's causing the issue now.

jkoritzinsky committed:
Add back FrameworkPackageName to get the partial platform manifest to generate.

ViktorHofer commented:
@jkoritzinsky is the PR ready I be merged?

jkoritzinsky commented:
Yep!

ViktorHofer commented:
Niceeee. I assume you will merge it after sending out the mail? :)

jkoritzinsky commented:
Will do!
